client/cmd.go

```diff
@@ -215,14 +216,26 @@ func ReadTxCommandFlags(clientCtx Context, flagSet *pflag.FlagSet) (Context, err
 		clientCtx = clientCtx.WithSkipConfirmation(skipConfirm)
 	}
 
+	if clientCtx.SignModeStr == "" || flagSet.Changed(flags.FlagSignMode) {
+		signModeStr, _ := flagSet.GetString(flags.FlagSignMode)
+		clientCtx = clientCtx.WithSignModeStr(signModeStr)
+	}
+
 	if clientCtx.From == "" || flagSet.Changed(flags.FlagFrom) {
 		from, _ := flagSet.GetString(flags.FlagFrom)
-		fromAddr, fromName, err := GetFromFields(clientCtx.Keyring, from, clientCtx.GenerateOnly)
+		fromAddr, keyringInfo, err := GetFromFields(clientCtx.Keyring, from, clientCtx.GenerateOnly)
 		if err != nil {
 			return clientCtx, err
 		}
 
-		clientCtx = clientCtx.WithFrom(from).WithFromAddress(fromAddr).WithFromName(fromName)
+		clientCtx = clientCtx.WithFrom(from).WithFromAddress(fromAddr).WithFromName(keyringInfo.GetName())
+
+		// If the `from` signer account is a ledger key, we need to use
+		// SIGN_MODE_AMINO_JSON, because ledger doesn't support proto yet.
+		// ref: https://github.com/cosmos/cosmos-sdk/issues/8109
+		if keyringInfo.GetType() == keyring.TypeLedger && clientCtx.SignModeStr != flags.SignModeLegacyAminoJSON {
+			clientCtx = clientCtx.WithSignModeStr(flags.SignModeLegacyAminoJSON)
```

Comment:
I could have just `flagSet.AddFlag("--sign-mode=amino-json")`  here, but it feels ugly, so I went with the clientCtx.WithSignModeStr dance.

Reply:
This looks preferred anyway